### PR TITLE
Add `Context` functionality

### PR DESCRIPTION
**Pull request by thomaseizinger**
Hi,

first of, thank you very much for this great library! It almost exactly fits our needs! I haven't done anything with custom derives before. so feedback to my changes is very much appreciated. I'll shortly explain what I am trying to achieve:

We want to be able to persist the states used by the state machine. To make this easier, we figured it would be great if the individual states would only store data, e.g. easily persistable structs. But this excludes storing the futures in there that actually drive the state machine.

The idea is to introduce a `Context` that you pass in as the state machine is started. Later on, you can be access the context in each poll method. I tried to outline the usage in the test I added.

So far it is working!

However, given I have never worked with custom derives before, I don't know about potential pitfalls when generating code. I am also very unsure if my changes are idiomatic or whether the needed functionality could be achieved in a simpler manner.

### CODE REVIEW

**Comment by bkchr** [comment]
First thanks for the pull request. I once thought about something similar, I often have state data that I need to pass between every state and it would be nice to just define this data once. 
So, this would fit nicely with our context proposal.

The first thing I would probably change would be to wrap the `Context` in a `RentToOwn`, to make it possible to consume that as well.

I will think further about this pull request and get back to you.

But nevertheless, thanks again :)

**Comment by thomaseizinger** [comment]
Thanks for the feedback!
I am not sure if consuming the Context should actually be possible since it is accessible in every state. If you consume it once, later states will not have access to the context.

Looking forward to further feedback! I'd welcome it if this functionality would be available in some way!

**Comment by thomaseizinger** [comment]
There is now actually an issue with `derive`s that are forwarded to the generated structs.

Currently, they are put on the generated structs for the states AND the created future.

However, with the proposed change, the `Future`-struct contains also the context, which is a type defined by the user. This conflicts with the custom derives now since it forces the user to derive everything they want to have on the states to be also present on the `Context`.

IMO, the derives should not be added to the `Future` struct. After all, there is not really anything you would want to do with this future other than spawning it in tokio. (Maybe the start method could also just make use of `impl Trait`?).

What do you think about that?

If you would want to keep the derives on the future, I think the way to go would be to make it configurable, which derives go on the states and which go on the generated `Future`.

**Comment by bkchr** [comment]
I'm stiltentoownl in favor of making `context` a `RentToOwn`. I'm aware that the `context` is gone, if it will be consumed in some state. We could add a check and ensure that the `context` is still valid, before we enter a state and otherwise `panic!`. Maybe you have a better idea for it? It would be nice, to make sure that if we consume the `context`, we must enter the `End` state.

Regarding the deriving stuff, selecting which traits are exclusive for the states would be the best idea in my opinion.

**Comment by thomaseizinger** [comment]
My thoughts are that I would like to avoid runtime errors like `panic!` if the `context` is not present.

Our use-case for `context` is for shared data the can be accessed in any state. For example, an `Arc` to a some kind of persistence interface, external service, etc: Long-lived, shared resources that are needed by the state machine in order to determine whether it should advance to the next state.

### Regarding `RentToOwn`:

I was quite fascinated when I first saw the construct - it is pretty neat and I think it makes perfect sense for the current state to be wrapped in that. Once you transition to the next one, the memory will be dropped anyway so you might as well take it.

Making the `context` a `RentToOwn` would not hurt our particular usecase as we only use `context` for things where you just need a reference anyway. I think the question is more if we would want users to being able to take it, for the cost of having to do runtime checks whether the `context` is still there.

Now that I am thinking about it, IIRC `RentToOwn` doesn't even provide a check to see if it is still there. It will just `panic!` inside the `unwrap` of the underlying `Option`. `context` being a `RentToOwn` would mean that users have to track themselves whether they have consumed it already.

Given the question if something should go into the state or the context, I'd say:

- Is it an actual _data_ structure? Would it make sense to implement traits like `PartialEq`, `Hash` or `Copy` on it? -> Put it into the state.
- Is it a shared pointer? (`Arc` or `Rc`?) -> Probably context.

### Regarding `derive` on the Future struct

I am happy to make it configurable but I'd question if the functionality is actually needed. Here is why:

Given the above "rule" of where certain data should go, _not_ adding the `derive`s to the `Future` struct would mean, only the `state` structs would get the derives. If the `states` only contain actual _data_ structs, adding `derive`s like `Serialize` or `PartialEq` is straight forward.

If, however, the `derive`s are added to the `Future` struct as well, the user would be forced to implement things like `Serialize` or `PartialEq` on a struct that only contains shared-pointers :man_shrugging:.
Not only is this kind of awkward but it actually raises the question of which `derive`s you would want to have on this `Future` or `Future`s in general?

The only one I can think of that sort of makes sense would be `Debug` but even that one is kind of weird: The instance of the `Future` is consumed as soon as you spawn it into the executor, when would you ever use that thing in a `log`/`println` statement so that the `Debug` implementation is used?

I checked some implementations of `Future` inside crates like `tokio`, `warp` or `hyper` and they actually never derive any traits. I guess the main reason for this is that `Future`s are usually either wrapped in other `Future`s or spawned into an executor because otherwise they don't provide value.

Long story short: For the sake of simplicity in the implementation and the configuration, I think the `derive` block should simply be omitted from the `Future` struct.

Maybe @fitzgen can add some insights on why it was initially added?

**Comment by thomaseizinger** [comment]
Ah yeah, fair point.
I didn't consider that using a RentToOwn actually requires the field to be an option. This in turn allows the state machine to determine whether the context is still there and just return NotReady without calling the function. As mentioned, it doesn't hurt our usecase so, happy to change that now that I understand the implications :D
Also, would mind outlining your usecase shortly? I'd be interested in what it looks like :)

**Comment by thomaseizinger** [comment]
I rebased onto current master and fixed the conflicts!

In addition, I wrapped the `context` instance in a `RentToOwn`. It was quite tricky because I had to nest the `RentToOwn::with` calls so that both variables, the current state and the context, are available.

As discussed, I removed the `derive` block from the actual future and added a `compile_fail` test that the derives are no longer there.

Last but not least, I a section about the context to the README that explains how it can be used. Through that, I actually discovered an edge-case in the code generation that I missed before.

Please take another look if we can merge it in this state and thanks for your input! :)

**Comment by bkchr** [comment]
Now you put the context in a `RentToOwn`, I would also accepted it without it, but okay :)

**Comment by thomaseizinger** [comment]
Yeah, as mentioned in my previous reply, I did not fully grasp what that would look like. Initially I thought that the state machine would just crash the next time it is polled if the context is taken once.

As you correctly mentioned, the user can already screw up the state machine by taking the state without advancing so I thought it is better to wrap it in a `RentToOwn` if this allows for even more use cases to be implemented.

**Comment by bkchr** [comment]
Ty :)
